proposals/0002-gep.md

```diff
@@ -0,0 +1,140 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# TITLE: Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+- R6: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R7: API routes must unify common LLM endpoints/API
+- R8: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: /chat/{pool-id}
+    complete: /complete/{pool-id} #this is also know as generate in some providers
+    multi-modal: /multimodal/{pool-id} # future route, not for MVP
+    embedding: /embedding/{pool-id} # future route, not for MVP
+```
+
+#### User Request Schema for Chat Route
+
+Below is an example user request to the Chat route. The other routes will look very similar, if not identical. The provider specific params will look different.
+
+Example Request Body:
+
+
+```json
+{
+  "message": "tell me a joke",
```

Comment:
I would love to see a bit more on the unified request schemas to chat/complete API like:
- how do the simplest possible requests look like for the unified chat and complete APIs? I don't think they are the same. In case of chat API, we need to provide chat history.
- how do we propose to unify parameters like top_p? Where they should be provided?
- What if I want to override a param or a few for a specific provider how would that look like? How would the request payload look like in that case?
- What if we want to override a prompt for a specific provider in Glide Chat API? How would we pass chat history in that case?

I would like to see more examples of payloads for these cases to understand the target Glide's API fully.

Reply:
how do we propose to unify parameters like top_p? Where they should be provided? - Isn't this provided via the config?

Reply:
I wonder if for the MVP we don't allow overrides. Everything gets updated via the config. I'm trying to keep the MVP simple so we can get feedback from users how these things should come together.

Reply:
@mkrueger12 
> how do we propose to unify parameters like top_p? Where they should be provided? - Isn't this provided via the config?

This meant to be unified in the chat API request schema in case users want to override some params on the fly. I'm good with leaving this out for now, but I would still plan for that in the request schamas, so we have some flexibility there (without that, schema extension in a backward-compatible way could be a teidius task :/ )